algorithms/kernel/svm/svm_predict_impl.i

```diff
@@ -67,57 +195,76 @@ struct SVMPredictImpl<defaultDense, algorithmFPType, cpu> : public Kernel
-        DAAL_CHECK_STATUS_VAR(s);
-
-        auto kfResultPtr = new kernel_function::Result();
-        DAAL_CHECK_MALLOC(kfResultPtr)
-        kernel_function::ResultPtr shRes(kfResultPtr);
-        shRes->set(kernel_function::values, shResNT);
-        kernel->setResult(shRes);
-        kernel->getInput()->set(kernel_function::X, xTable);
-        kernel->getInput()->set(kernel_function::Y, svTable);
-        kernel->getParameter()->computationMode = kernel_function::matrixMatrix;
-        s                                       = kernel->computeNoThrow();
-        if (!s) return Status(services::ErrorSVMPredictKernerFunctionCall).add(s); //this order is expected by test system
-
-        char trans  = 'T';
-        DAAL_INT m_ = nSV;
-        DAAL_INT n_ = nVectors;
-        algorithmFPType alpha(1.0);
-        DAAL_INT lda = m_;
-        DAAL_INT incx(1);
-        algorithmFPType beta(1.0);
-        DAAL_INT incy(1);
-
-        service_memset<algorithmFPType, cpu>(distance, bias, nVectors);
-
-        Blas<algorithmFPType, cpu>::xgemv(&trans, &m_, &n_, &alpha, buf, &lda, svCoeff, &incx, &beta, distance, &incy);
-
-        return s;
+        size_t nRowsPerBlock = 1;
+        DAAL_SAFE_CPU_CALL((nRowsPerBlock = 256), (nRowsPerBlock = nVectors));
+
+        const size_t nBlocks = nVectors / nRowsPerBlock + !!(nVectors % nRowsPerBlock);
+
+        /* TLS data initialization */
+        daal::tls<PredictTask<algorithmFPType, cpu> *> tlsTask([&]() {
+            if (xTable->getDataLayout() == NumericTableIface::csrArray)
+            {
+                return PredictTaskCSR<algorithmFPType, cpu>::create(nRowsPerBlock, xTable, svTable, kernel);
+            }
+            else
+            {
+                return PredictTaskDense<algorithmFPType, cpu>::create(nRowsPerBlock, xTable, svTable, kernel);
+            }
+        });
+
+        SafeStatus safeStat;
+
+        daal::threader_for(nBlocks, nBlocks, [&](int iBlock) {
+            PredictTask<algorithmFPType, cpu> * tlsLocal = tlsTask.local();
+
+            const size_t startRow          = iBlock * nRowsPerBlock;
+            const size_t endRow            = services::internal::min<cpu, size_t>(startRow + nRowsPerBlock, nVectors);
+            const size_t nRowsPerBlockReal = endRow - startRow;
+
+            DAAL_CHECK_THR(tlsLocal->kernelCompute(startRow, nRowsPerBlockReal), services::ErrorSVMPredictKernerFunctionCall);
+
+            const algorithmFPType * buf = tlsLocal->getBuff();
+
+            char trans  = 'T';
+            DAAL_INT m_ = nSV;
+            DAAL_INT n_ = nRowsPerBlockReal;
+            algorithmFPType alpha(1.0);
+            DAAL_INT lda = m_;
+            DAAL_INT incx(1);
+            algorithmFPType beta(1.0);
+            DAAL_INT incy(1);
+
+            if (nBlocks == 1)
+            {
+                Blas<algorithmFPType, cpu>::xgemv(&trans, &m_, &n_, &alpha, buf, &lda, svCoeff, &incx, &beta, distance + startRow, &incy);
```

Comment:
why we need threaded gemv version here inside threader_for?

Reply:
Yes, for `nBlock = 1`. This means that there will be one thread.